docs/xdsl-introduction.ipynb

```diff
@@ -0,0 +1,937 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "81ccec58-8c70-4a47-b167-558ae3903bfb",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "# The Multi-Levelled Approach to Compilation\n",
+    "\n",
+    "Most programming languages are rather far removed from Assembly code. As a result, translating a language directly into Assembly is both difficult (easy to make mistakes) and inefficient (generated Assembly code is slow!). Therefore, most compilers goes through many intermediate representations / languages (IRs), which represents different layers of abstractions, to provide an efficient translation down to Assembly code - which is also feasible for compiler engineers to write. \n",
+    "\n",
+    "Although an effective approach, over the years it has become apparent that many compilers are repeatedly defining similar intermediate languages. Given the separate architecture of different compilers, it has been quite difficult to reuse code between architectures. \n",
+    "\n",
+    "Previous projects such as LLVM has provided a common abstraction layer for different compilers to reuse - however this is only one layer, and compilers of different domains are still having to repeat significant amounts of work to reach the LLVM level. \n",
+    "\n",
+    "LLVM is also not necessarily the ideal layer to perform optimisations on - thus many compilers will also implement their own implementations on custom IRs before they reach LLVM.\n",
+    "\n",
+    "xDSL is a frameworks which make it easy to:\n",
+    "1. Define custom abstraction layers for compilers\n",
+    "2. Reuse abstractions from other compilers\n",
+    "3. Encode domain-specific knowledge (i.e. optimisations) within these abstraction layers\n",
+    "\n",
+    "xDSL takes inspriation from MLIR, which is a framework based upon similar ideas, and is implemented in C++.\n",
+    "\n",
+    "<!-- Covered here: --!>\n",
+    "<!-- - How compilers use IRs --!>\n",
+    "<!-- - Why a framework for IRs are needed and is useful (ease of specifying optimisations)  --!>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6f22e5fe-8f6c-4da5-9f82-b1a39d21abcc",
+   "metadata": {},
+   "source": [
+    "# General Concepts\n",
+    "\n",
+    "The kind of IRs that xDSL builds are SSA-based - a restriction placed on IRs which makes it easy for analysis of the code."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ca733af2-dd0f-400f-9b69-de1878d69443",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "### Static Single Assignment (SSA) \n",
+    "The kind of programming languages xDSL defines have the $\\textit{static single assignment}$ (SSA) property. It is a property on variables that means they can be assigned a value only once, and once assigned a value, they cannot be modified. This is a common restriction used within compilers, as it enables many compiler optimisations to be performed.\n",
```

Comment:
I think that speaking of SSA programming languages here is confusing. The IRs are SSA, but this sentence could be read as "xDSL can't be used to compile non-SSA languages".
Generally, my main review would be to use "IR" and "programming language" less interchangeably throughout the notebook. I think it is confusion-prone, especially for an introduction. The exact distinction is an interesting discussion in itself, but I think it is useful to differentiate them at this point.
What do others think?

Reply:
good point & thanks for noticing,  i think its a good idea to settle on one as well
curious to hear what others think too!